### PR TITLE
NeVa::forward - remove device syncs (torch.where) and vectorize over batch dimensions

### PR DESCRIPTION
**Pull request by nikitaved**
The following patch modifies the forward of the NeVa model (the `replace_media_embeddings` method) by introducing the following improvements:

- Elimination of `torch.where` with a single argument. That call introduces device synchronizations when run on GPUs.
- Media index selection manipulation in the construction of `padded_media_indices` can be done  in a single vectorized call.
- Reduction in the overall number of kernel launches.
- Elimination of calls to `all`. This also causes device syncs. We can leave it, or run it in debug mode only. It is best to have a dedicated test instead of that assert, imho.

### CODE REVIEW

File: nemo/collections/multimodal/models/multimodal_llm/neva/neva_model.py [hunk 1]
```diff
@@ -232,22 +232,32 @@ def replace_media_embeddings(self, input_ids, inputs_embeds, media):
...
+            )
+        else:
+            padded_media_indices = torch.where(
+                sorted_media_end_positions_mask.to(torch.bool),
+                media_end_positions_mask_sort_idx - num_patches + 1,
+                sequence_length
+            )
+
+        # TODO: add asserts that were here below
+        # }
```
Comment (nikitaved) [inline nemo/collections/multimodal/models/multimodal_llm/neva/neva_model.py:262]: The assert I mean the call to `all`. These also cause device synchronizations. They are better be reserved for runs in a some kind of debug mode. It is best to have a dedicated test though. Where would one place such a test?

Comment (yaoyu-33) [reply nemo/collections/multimodal/models/multimodal_llm/neva/neva_model.py:262]: @nikitaved  can you add condition if env var `NEMO_TESTING` is True, still run assert?

Comment (nikitaved) [reply nemo/collections/multimodal/models/multimodal_llm/neva/neva_model.py]: @yaoyu-33 , done! Is there a way to get some CI output?

**Comment by nikitaved** [comment]
@yaoyu-33 , could you please have another look and let the workflow be run?

**Comment by tfogal** [comment]
FWIW, I ran our little test (a `neva_pretrain.py` invocation) with `NEMO_TESTING=1` and things ran fine there.

**Review by yaoyu-33 — APPROVED** [review]

**Comment by nikitaved** [comment]
@yaoyu-33 , @ericharper , could you please run the workflow again? Thank you!

**Comment by tfogal** [comment]
bump to kill 'stale'.
i checked that i have access to an internal cluster today, will check convergence soon

**Comment by nikitaved** [comment]
Thanks you, @tfogal ! FWIW, the very last CI run turned out to be green

**Comment by nikitaved** [comment]
Is it possible to move it forward? Commenting to remove the current stale status.

**Comment by tfogal** [comment]
> Remove stale label or comment or update or this will be closed in 7 days.

comment

**Comment by nikitaved** [comment]
"
Wait it out
Gonna wait it out
Wait it out (be patient)
" Patient, Tool

**Comment by nikitaved** [comment]
Is there interest in reviving this PR? I assume it is rather low impact with perf benefits assuming the tests are green. And green they were the last time CIs were run.
